### PR TITLE
[IMP] mail: turn "RtcSession/rtcDataChannel" into field

### PR DESCRIPTION
**Pull request by alexkuhn**
Task-2734101

### CODE REVIEW

File: addons/mail/static/src/models/rtc/rtc.js [hunk 1]
```diff
@@ -512,7 +504,12 @@ registerModel({
                 }
             };
             this._peerConnections[token] = peerConnection;
-            this._dataChannels[token] = dataChannel;
+            this.update({
+                dataChannels: {
+                    ...this.dataChannels,
+                    [token]: dataChannel,
+                },
+            });
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc/rtc.js:512]: make model instead of having object field

File: addons/mail/static/src/models/rtc/rtc.js [hunk 2]
```diff
@@ -746,11 +738,10 @@ registerModel({
             if (rtcSession) {
                 rtcSession.reset();
             }
-            const dataChannel = this._dataChannels[token];
-            if (dataChannel) {
-                dataChannel.close();
+            const rtcDataChannel = this.messaging.models['RtcDataChannels'].findFromIdentifyingData({ token });
+            if (rtcDataChannel) {
+                rtcDataChannel.delete();
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc/rtc.js:743]: Not for this PR, but when we have all rtc models, it looks like data channel should be more closely related to peer (probably with causality).

File: addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    name: 'RtcDataChannel',
+    identifyingFields: ['token'],
+    lifecycleHooks: {
+        _willDelete() {
+            this.dataChannel.close();
+        },
+    },
+    fields: {
+        dataChannel: attr(),
+        token: attr(),
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js:16]: both probably need to be required and readonly

File: addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+registerModel({
+    name: 'RtcDataChannel',
+    identifyingFields: ['token'],
+    lifecycleHooks: {
+        _willDelete() {
+            this.dataChannel.close();
+        },
+    },
+    fields: {
+        dataChannel: attr(),
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js:15]: ideally `dataChannel` should be auto-created with a compute

File: addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+/** @odoo-module **/
+
+import { registerModel } from '@mail/model/model_core';
+import { attr } from '@mail/model/model_field';
+
+registerModel({
+    name: 'RtcDataChannel',
+    identifyingFields: ['token'],
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js:8]: same comment as in other PRs, should probably have session record relation as identifying field

Comment (seb-odoo) [reply addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js:8]: I would still much prefer if we cleaned this up before merging, to use proper identifying field (I think it should be a one2one with rtc session, which is the holder of the token).

Comment (alexkuhn) [reply addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js:8]: I can't just define a `one2one` with `RtcSession` with current code, because it makes use of many `dataChannels` while there's currently just a `one` with `Rtc/currentRtcSession`.

Are you sure `RtcDataChannel/token` is conceptually the same as `RtcSession/id`?
- If yes, I could do some changes to go in direction of what you say, like defining `Rtc/rtcSessions` that is a `many` of all known `RtcSession`s.
- If not sure, then it's much safer to keep mode isolated for the time being, and reconcile at a later date when the code will make far more sense when they are conceptually the same.

Comment (ThanhDodeurOdoo) [reply addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js]: Yes, `token`/`peerToken` is `RtcSession/id`, It a remnant from when we used channel members instead of "temporary" rtc sessions, I should have refactored it.

File: addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        },
+    },
+    recordMethods: {
+        /**
+         * @private
+         * @returns {FieldCommand}
+         */
+        _computeRtc() {
+            return this.messaging && this.messaging.rtc ? replace(this.messaging.rtc) : clear();
+        },
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js:22]: This is just a related?
And why do we need to define the field specifically?

File: addons/mail/static/src/models/rtc/rtc.js [hunk 7]
```diff
@@ -226,6 +217,7 @@ registerModel({
                 currentRtcSession: clear(),
                 disconnectAudioMonitor: clear(),
                 logs: clear(),
+                rtcDataChannels: clear(),
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc/rtc.js:220]: I don't think we want this field on rtc, this should be automatically done when deleting the related sessions.

File: addons/mail/static/src/models/rtc/rtc.js [hunk 8]
```diff
@@ -513,7 +505,10 @@ registerModel({
                 }
             };
             this._peerConnections[token] = peerConnection;
-            this._dataChannels[token] = dataChannel;
+            this.messaging.models['RtcDataChannels'].insert({
+                dataChannel,
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc/rtc.js:509]: This is done in reverse, the RtcDataChannels model should create the actual channel, not the other way around.
Ok to do in another PR if that allows reducing diff here.

Comment (seb-odoo) [reply addons/mail/static/src/models/rtc/rtc.js:509]: Also model name should be singular

File: addons/mail/static/src/models/rtc/rtc.js [hunk 9]
```diff
@@ -673,11 +668,15 @@ registerModel({
             }
             if (type === 'peerToPeer') {
                 for (const token of targetTokens) {
-                    const dataChannel = this._dataChannels[token];
-                    if (!dataChannel || dataChannel.readyState !== 'open') {
+                    const rtcSession = this.messaging.models['RtcSession'].findFromIdentifyingData({ id: token });
+                    if (!rtcSession) {
+                        continue;
+                    }
+                    const rtcDataChannel = rtcSession.rtcDataChannels;
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc/rtc.js:675]: wrong field name

File: addons/mail/static/src/models/rtc/rtc.js [hunk 10]
```diff
@@ -1300,6 +1298,12 @@ registerModel({
         recoveryDelay: attr({
             default: 3000,
         }),
+        /**
+         * Contains the RTCDataChannels with the other rtc sessions.
+         */
+        rtcDataChannels: many('RtcDataChannel', {
+            inverse: 'rtc',
+        }),
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc/rtc.js:1306]: probably just remove this from here

File: addons/mail/static/src/models/rtc/rtc.js (hunk 7 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/rtc/rtc.js:220]: there is no `clear()` of all `RtcSession`s here. Only the `currentRtcSession`.
So I'm not sure whether removing this LOC is safe.

File: addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js (hunk 6 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js:22]: Was done so that we could delete all `RtcDataChannel`s in specific flow of `Rtc`.
I removed this and instead just loop on all `RtcDataChannel` and delete them, without relying on a relational field.

File: addons/mail/static/src/models/rtc/rtc.js (hunk 8 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/rtc/rtc.js:509]: Let's do this in another PR. Keep focus on just turning props in fields with this PR and others.
At a glance, create `dataChannel` in `RtcDataChannel` is more refactoring. `_createPeerConnection` is quite big and lots of call to other methods in `Rtc`.
And also `_peerConnections` has the same issues, so this would lead to big PR if we turn into fields and make this change.

File: addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js (hunk 5 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/rtc_data_channel/rtc_data_channel.js]: > Yes, `token`/`peerToken` is `RtcSession/id`, It a remnant from when we used channel members instead of "temporary" rtc sessions, I should have refactored it.

Cool, so will refactor so that all `token`/`peerToken` dicts are instead record that are identified by the `RtcSession` (since `RtcSession/id` is identifying field of `RtcSession`).

**Review by phenix-factory — APPROVED** [review]
LGTM

File: addons/mail/static/src/models/rtc.js [hunk 11]
```diff
@@ -754,11 +755,10 @@ registerModel({
             if (rtcSession) {
                 rtcSession.reset();
             }
-            const dataChannel = this._dataChannels[sessionId];
-            if (dataChannel) {
-                dataChannel.close();
+            const rtcDataChannel = rtcSession.rtcDataChannel;
```
Comment (ThanhDodeurOdoo) [inline addons/mail/static/src/models/rtc.js:758]: `rtcSession` can already be deleted in some asynchonous calls to `_removePeer()`, that's why it's guarded above, maybe this part should also be in the guard to avoid tracebacks when trying to read `rtcDataChannel` of `undefined`.

Comment (alexkuhn) [reply addons/mail/static/src/models/rtc.js:758]: `delete()` is on `RtcDataChannel`: it changes the modelling so that there is no more `RtcDataChannel` related to this `RtcSession`. `close()` is on `RtcDataChannel/dataChannel`, and it's auto-closed on deletion of the `RtcDataChannel` (see `RtcDataChannel/_willDelete`).

**Comment by alexkuhn** [comment]
@robodoo r+
